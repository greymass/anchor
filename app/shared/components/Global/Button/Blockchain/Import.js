// @flow
import React, { Component } from 'react';
import { translate } from 'react-i18next';
import { Button } from 'semantic-ui-react';

import GlobalModalBlockchainImport from '../../Modal/Blockchain/Import';

export class GlobalButtonBlockchainImport extends Component<Props> {
  state = {
    open: false
  };

  onOpen = () => this.setState({ open: true });
  onClose = () => this.setState({ open: false });

  render() {
    const {
      connection,
      settings,
      t
    } = this.props;
    const {
      open
    } = this.state;
    return (
      <GlobalModalBlockchainImport
        connection={connection}
        onClose={this.onClose}
        open={open}
        settings={settings}
        trigger={(
          <Button
            color="blue"
            content={t('global_button_blockchain_import_action')}
            icon="circle plus"
            onClick={this.onOpen}
          />
        )}
      />
    );
  }
}

export default translate('global')(GlobalButtonBlockchainImport);
