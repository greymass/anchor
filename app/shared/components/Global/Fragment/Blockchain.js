// @flow
import React, { Component } from 'react';
import { Popup } from 'semantic-ui-react';
import { translate } from 'react-i18next';

export class GlobalFragmentBlockchain extends Component<Props> {
  render() {
    const {
      blockchain,
      t
    } = this.props;
    if (blockchain) {
      return (
        <Popup
          content={`Chain ID: ${blockchain.chainId}`}
          hoverable
          inverted
          trigger={(
            <span>
              {blockchain.name}
            </span>
          )}
        />
      );
    }
    return false;
  }
}

export default translate('global')(GlobalFragmentBlockchain);
