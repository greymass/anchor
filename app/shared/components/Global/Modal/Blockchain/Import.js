// @flow
import React, { Component } from 'react';
import { translate } from 'react-i18next';
import { Header, Modal } from 'semantic-ui-react';

import GlobalBlockchainImport from '../../../../containers/Global/Blockchain/Import';

export class GlobalModalBlockchainImport extends Component<Props> {
  render() {
    const {
      onClose,
      open,
      t,
      trigger
    } = this.props;
    return (
      <Modal
        centered={false}
        trigger={trigger}
        onClose={onClose}
        open={open}
        size="small"
      >
        <Header
          content={t('global_button_blockchain_import_action')}
          icon="users"
          subheader={t('global_button_blockchain_import_action')}
        />
        <GlobalBlockchainImport onClose={onClose} />
      </Modal>
    );
  }
}

export default translate('global')(GlobalModalBlockchainImport);
