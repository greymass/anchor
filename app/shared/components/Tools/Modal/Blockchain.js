// @flow
import React, { Component } from 'react';
import { translate } from 'react-i18next';
import { Header, Modal } from 'semantic-ui-react';

import GlobalBlockchainForm from '../../../containers/Global/Blockchain/Form';

class ToolsModalBlockchain extends Component<Props> {
  render() {
    const {
      blockchain,
      onClose,
      onSuccess,
      open,
      t,
      trigger
    } = this.props;

    return (
      <Modal
        centered={false}
        trigger={trigger}
        onClose={onClose}
        open={open}
      >
        <Header icon="cubes" content={t('tools_modal_blockchains_header_text')} />
        <Modal.Content>
          <GlobalBlockchainForm
            blockchain={blockchain}
            onCancel={onClose}
            onSubmit={onSuccess}
          />
        </Modal.Content>
      </Modal>
    );
  }
}

export default translate('tools')(ToolsModalBlockchain);
