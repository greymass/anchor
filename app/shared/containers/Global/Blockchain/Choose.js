// @flow
import React, { Component } from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';

import { withRouter } from 'react-router-dom';
import { withTranslation } from 'react-i18next';
import compose from 'lodash/fp/compose';

import { Card, Form, Header, Segment, Step } from 'semantic-ui-react';

import GlobalFragmentChainLogo from '../../../components/Global/Fragment/ChainLogo';
import { swapBlockchain } from '../../../actions/blockchains';
import { setSetting } from '../../../actions/settings';

import GlobalBlockchainDropdown from './Dropdown';

const featuredChains = [
  // EOS
  'aca376f206b8fc25a6ed44dbdc66547c36c6c33e3a119ffbeaef943642f0e906',
  // FIO
  '21dcae42c0182200e93f954a074011f9048a7624c6fe81d3c9541a614a88bd1c',
  // TELOS
  '4667b205c6838ef70ff7988f6e8257e8be0e1284a2f59699054a018f743b1d11',
  // PROTON
  '384da888112027f0321850a169f737c33e53b388aad48b5adace4bab97f437e0',
  // WAX
  '1064487b3cd1a897ce03ae5b6a865651747e2e152090f99c1d19d44e01aea5a4',
];

class GlobalBlockchainChoose extends Component<Props> {
  state = {
    showTestnets: false
  }
  onChange = (e, { checked }) => this.setState({ showTestnets: checked })
  onSwap = (chainId) => {
    const {
      actions,
      history,
    } = this.props;
    actions.swapBlockchain(chainId);
    actions.setSetting('blockchains', [chainId]);
    history.push('');
  }
  render() {
    const {
      blockchains,
      t,
    } = this.props;
    const {
      showTestnets
    } = this.state;
    const featured = blockchains.filter(b => featuredChains.includes(b.chainId));
    return (
      <Segment style={{ marginTop: 0 }}>
        <Header
          content={t('global_blockchain_choose_header')}
          subheader={t('global_blockchain_choose_subheader')}
          size="large"
          style={{ marginTop: 0 }}
        />
        <Segment basic padded="very" textAlign="center">
          <Card.Group style={{
              'align-items': 'center',
              'justify-content': 'center',
            }}
          >
            {featured.map(blockchain => (
              <Card
                onClick={() => this.onSwap(blockchain.chainId)}
                raised
                style={{ width: 'auto' }}
              >
                <GlobalFragmentChainLogo
                  chainId={blockchain.chainId}
                  noPopup
                  style={{
                    width: '96px'
                  }}
                />
                <Card.Content>
                  <Card.Header>{blockchain.name}</Card.Header>
                </Card.Content>
              </Card>
            ))}
          </Card.Group>
          <Form style={{ marginTop: '3em' }}>
            <Header size="small">
              More options available in the dropdown below.
            </Header>
            <Form.Field>
              <GlobalBlockchainDropdown
                initialize
                onSwap={this.onSwap}
                search
                selection
                showTestnets={showTestnets}
                style={{ width: '30em' }}
              />
            </Form.Field>
            <Form.Field>
              <Form.Checkbox
                label={t('global_blockchain_choose_show_testnets')}
                onChange={this.onChange}
              />
            </Form.Field>
          </Form>
        </Segment>
      </Segment>
    );
  }
}

function mapStateToProps(state) {
  return {
    blockchains: state.blockchains,
    navigation: state.navigation,
    settings: state.settings,
    wallets: state.wallets
  };
}

function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators({
      setSetting,
      swapBlockchain,
    }, dispatch)
  };
}

export default compose(
  withTranslation('global'),
  withRouter,
  connect(mapStateToProps, mapDispatchToProps)
)(GlobalBlockchainChoose);
