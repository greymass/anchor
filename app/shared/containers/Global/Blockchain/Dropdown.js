// @flow
import React, { Component } from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { withTranslation } from 'react-i18next';
import compose from 'lodash/fp/compose';
import { find } from 'lodash';
import { Button, Dropdown, Header, Label, Segment } from 'semantic-ui-react';

import GlobalFragmentChainLogo from '../../../components/Global/Fragment/ChainLogo';
import * as BlockchainsActions from '../../../actions/blockchains';

class GlobalBlockchainDropdown extends Component<Props> {
  swapBlockchain = (chainId) => {
    const { actions } = this.props;
    actions.swapBlockchain(chainId);
    if (this.props.onSwap) {
      this.props.onSwap(chainId);
    }
  };
  render() {
    const {
      blockchains,
      chainIds,
      disabled,
      fluid,
      initialize,
      onSelect,
      selected,
      selection,
      settings,
      showName,
      showTestnets,
      style,
      t,
    } = this.props;
    if (
      !initialize
      && (
        !settings.blockchains
        || settings.blockchains.length === 0
        || !settings.walletInit
      )
    ) {
      return false;
    }
    const { chainId } = settings;
    if (!blockchains) return false;
    let blockchain = find(blockchains, { chainId });
    if (!blockchain) {
      blockchain = {};
    }
    if (selected) {
      blockchain = find(blockchains, { chainId: selected });
    }
    const options = blockchains
      .filter(b => {
        const { pinnedBlockchains } = settings;
        const hasPins = (pinnedBlockchains && pinnedBlockchains.length > 0);
        const isEnabled = settings.blockchains.includes(b.chainId);
        const isCurrent = b.chainId === blockchain.chainId;
        const isPinned = (hasPins && pinnedBlockchains.includes(b.chainId));
        const isSelectable = (!chainIds || chainIds.includes(b.chainId));
        // Return all chains while initializing
        if (
          initialize
          && (!b.testnet || showTestnets)
        ) return true;
        // Otherwies return based on user preference
        return (
          !isCurrent
          && isEnabled
          && isSelectable
          && (
            !hasPins
            || isPinned
          )
        );
      })
      .sort((a, b) => a.name > b.name)
      .map((b) => {
        const image = <GlobalFragmentChainLogo avatar chainId={b.chainId} name={b.name} />;
        return {
          props: {
            image,
            key: b.chainId,
            onClick: () => ((onSelect) ? onSelect(b.chainId) : this.swapBlockchain(b.chainId)),
            text: b.name,
            value: b.chainId,
          },
          b
        };
      });
    const trigger = (
      <span>
        {(blockchain && blockchain.chainId)
          ? (
            <GlobalFragmentChainLogo
              chainId={blockchain.chainId}
              noPopup
              style={{
                float: 'left',
                height: '2em',
                width: '2em',
              }}
            />
          )
          : t('dropdown_select_blockchain')
        }
        {(blockchain.testnet && !chainIds)
          ? (
            <Label color="orange" content="Testnet" size="large" />
          )
          : false
        }
        {(showName && blockchain && blockchain.name)
          ? (
            <Header
              content={blockchain.name}
              subheader={`${blockchain.chainId.substr(0, 6)}...${blockchain.chainId.substr(-6)}`}
              size="small"
              style={{ margin: 0 }}
            />
          )
          : false
        }
      </span>
    );
    if (disabled) {
      return (
        <Segment
          content={trigger}
          style={{ margin: 0 }}
        />
      );
    }
    const content = (
      <Dropdown.Menu key="menu" scrolling={!selection} style={{ minWidth: '200px', marginTop: 0 }}>
        {(this.props.onNavigationChange)
          ? (
            <Dropdown.Header>
              <Button
                basic
                content={t('button_manage_blockchains')}
                fluid
                icon="cubes"
                onClick={() => this.props.onNavigationChange('manage/blockchains')}
                size="small"
              />
            </Dropdown.Header>
          )
          : false
        }
        {options.map(option => {
          const {
            props,
          } = option;
          return (
            <Dropdown.Item {...props} />
          );
        })
        }
      </Dropdown.Menu>
    );
    const wrapper = (
      <Dropdown.Menu key="parent">
        {content}
      </Dropdown.Menu>
    );
    return (
      <Dropdown
        fluid={fluid}
        item
        labeled
        placeholder={showName ? t('dropdown_select_blockchain') : false}
        selection={selection}
        style={style}
        trigger={trigger}
      >
        {(selection)
          ? content
          : wrapper
        }
      </Dropdown>
    );
  }
}


function mapStateToProps(state) {
  return {
    blockchains: state.blockchains,
    settings: state.settings,
    validate: state.validate,
    wallet: state.wallet,
    wallets: state.wallets
  };
}

function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators({
      ...BlockchainsActions,
      // ...WalletActions,
      // ...WalletsActions,
    }, dispatch)
  };
}

export default compose(
  withTranslation('menu'),
  connect(mapStateToProps, mapDispatchToProps)
)(GlobalBlockchainDropdown);
