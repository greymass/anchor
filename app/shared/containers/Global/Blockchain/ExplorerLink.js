// @flow
import React, { PureComponent } from 'react';
import { connect } from 'react-redux';
import { find } from 'lodash';

import ExplorerLink from '../../../../shared/components/Global/Modal/ExplorerLink';

class GlobalBlockchainExplorerLink extends PureComponent<Props> {
  render = () => (
    <ExplorerLink
      blockExplorers={this.props.blockExplorers}
      content={this.props.content}
      linkData={this.props.linkData}
      linkType={this.props.linkType}
      settings={this.props.settings}
    />
  );
}

function mapStateToProps(state, ownProps) {
  const chainId = (ownProps.chainId) ? ownProps.chainId : state.settings.chainId;
  const blockchain = find(state.blockchains, { chainId });
  return {
    blockchain,
    blockExplorers: state.blockexplorers[blockchain._id],
    settings: state.settings,
  };
}

export default connect(mapStateToProps)(GlobalBlockchainExplorerLink);
