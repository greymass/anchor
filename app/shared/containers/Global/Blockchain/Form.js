// @flow
import React, { Component } from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import compose from 'lodash/fp/compose';
import { withTranslation } from 'react-i18next';
import { Accordion, Button, Divider, Form, Icon, Message } from 'semantic-ui-react';
import { find } from 'lodash';
import { delete as del, set } from 'dot-prop-immutable';

import GlobalFormFieldString from '../../../components/Global/Form/Field/String';
import GlobalFormFieldServer from '../../../components/Global/Form/Field/Url';

import BlockchainsActions from '../../../actions/blockchains';
import SettingsActions from '../../../actions/settings';
import ValidateActions from '../../../actions/validate';

class GlobalBlockchainForm extends Component<Props> {
  constructor(props) {
    super(props);
    this.state = {
      errors: {},
      loading: {},
      newChain: !!(props.chainId === 'new'),
      showAdvanced: !!(props.chainId === 'new'),
      values: Object.assign({}, props.blockchain),
      valids: {
        endpoint: true //props.settings.walletMode === 'cold' || false
      },
    };
  }
  static getDerivedStateFromProps(props, state) {
    const { validate } = props;
    let derived = {};
    if (state.values.node) {
      if (validate.NODE === 'PENDING') {
        // derived = set(state, 'valids.endpoint', false);
        // derived = set(derived, 'loading.endpoint', true);
        derived = set(state, 'errors', {});
      }
      if (validate.NODE === 'FAILURE') {
        // derived = set(state, 'errors.endpoint', 'undefined');
        derived = del(derived, 'loading.endpoint');
      }
      if (validate.NODE === 'SUCCESS') {
        derived = set(state, 'errors', {});
        // derived = set(state, 'valids.endpoint', true);
        // derived = del(derived, 'loading.endpoint');
      }
    }
    return derived;
  }
  componentDidMount() {
    this.props.actions.clearValidationState();
    const {
      settings
    } = this.props;
    const {
      chainId,
      node
    } = this.state.values;
    if (settings.walletMode !== 'cold' && node && chainId) {
      // this.props.actions.validateNode(node, chainId);
    }
  }
  showAdvanced = () => this.setState({ showAdvanced: !this.state.showAdvanced });
  isValid = () => !Object.values(this.state.valids).includes(false)
  onCancel = () => {
    this.state = {
      values: Object.assign({}, this.props.blockchain)
    };
    const { onCancel } = this.props;
    if (onCancel) {
      onCancel();
    }
  }
  onChange = (e, { name, valid, value }) => this.setState({
    values: set(this.state.values, name, value),
    valids: set(this.state.valids, name, valid),
  })
  onNodeChange = (e, { name, valid, value }) => this.setState({
    values: set(this.state.values, name, value),
    valids: set(this.state.valids, name, valid),
  }, () => {
    // this.props.actions.validateNode(value, this.state.values.chainId, false)
  })
  onSelect = (e, { name, value }) => this.setState({
    values: set(this.state.values, name, value)
  })
  onCheck = (e, { name, checked }) => this.setState({
    values: set(this.state.values, name, checked)
  })
  onKeyPress = (e) => {
    if (e.key === 'Enter') {
      this.onSubmit(e);
      e.preventDefault();
      return false;
    }
  }
  onSubmit = () => {
    const {
      actions,
      onCancel,
      onSubmit,
      settings
    } = this.props;
    const { values } = this.state;
    if (this.isValid()) {
      const isCurrentChain = !!(settings.chainId === values.chainId);
      if (isCurrentChain) {
        actions.setSetting('node', values.node);
      }
      actions.updateBlockchain(values);
      if (onSubmit) {
        onSubmit();
      }
      if (onCancel) {
        onCancel();
      }
    }
  };
  render() {
    const {
      app,
      settings,
      t,
    } = this.props;
    const {
      errors,
      loading,
      newChain,
      showAdvanced,
      values,
    } = this.state;
    const options = app.features.map((feature) => ({
      key: feature,
      text: t(`global:global_app_feature_${feature}_name`),
      value: feature,
    }));
    const {
      chainId,
      keyPrefix,
      name,
      node,
      supportedContracts,
      symbol,
      testnet,
    } = values;
    const hasErrors = (Object.keys(errors).length > 0);
    return (
      <Form
        error={hasErrors}
        onKeyPress={this.onKeyPress}
        onSubmit={this.onSubmit}
      >
        <GlobalFormFieldString
          autoFocus={newChain}
          defaultValue={chainId || ''}
          disabled={!newChain}
          label={t('tools_form_blockchain_chainid_label')}
          name="chainId"
          onChange={this.onChange}
        />
        <GlobalFormFieldString
          autoFocus={!newChain}
          defaultValue={name || ''}
          label={t('tools_form_blockchain_name_label')}
          name="name"
          onChange={this.onChange}
        />
        {(settings.walletMode !== 'cold')
          ? (
            <GlobalFormFieldServer
              defaultValue={node || ''}
              label={t('tools_form_blockchain_node_label')}
              loading={loading.endpoint}
              name="node"
              onChange={this.onNodeChange}
            />
          )
          : false
        }
        <Form.Field>
          <Accordion fluid styled>
            <Accordion.Title active={showAdvanced} index={0} onClick={this.showAdvanced}>
              <Icon name='dropdown' />
              {t('tools_form_blockchain_advanced_configuration')}
            </Accordion.Title>
            <Accordion.Content active={showAdvanced}>
              <GlobalFormFieldString
                defaultValue={keyPrefix || ''}
                label={t('tools_form_blockchain_keyprefix_label')}
                name="keyPrefix"
                onChange={this.onChange}
              />
              <GlobalFormFieldString
                defaultValue={symbol || ''}
                label={t('tools_form_blockchain_symbol_label')}
                name="symbol"
                onChange={this.onChange}
              />
              <Form.Select
                defaultValue={supportedContracts}
                fluid
                label={t('tools_form_blockchain_features_label')}
                multiple
                name="supportedContracts"
                onChange={this.onSelect}
                options={options}
                placeholder={t('tools_form_blockchain_features_placeholder')}
                selection
              />
            </Accordion.Content>
          </Accordion>
        </Form.Field>
        <Form.Field>
          <Form.Checkbox
            checked={testnet}
            label={t('tools_form_blockchain_testnet_label')}
            name="testnet"
            onChange={this.onCheck}
          />
        </Form.Field>
        <Divider hidden />
        {(hasErrors)
          ? (
            <Message error>
              {Object.keys(errors).map((error) => (
                <p>{t(`tools_form_blockchain_node_error_${errors[error]}`)}</p>
              ))}
            </Message>
          )
          : false
        }
        <Button
          content={t('cancel')}
          onClick={this.onCancel}
          secondary
        />
        <Button
          content={t('save')}
          disabled={(!this.isValid() || hasErrors)}
          icon="save"
          floated="right"
          primary
        />
      </Form>
    );
  }
}

function mapStateToProps(state, ownProps) {
  return {
    app: state.app,
    blockchain: find(state.blockchains, { chainId: ownProps.blockchain }),
    chainId: ownProps.blockchain,
    settings: state.settings,
    validate: state.validate
  };
}

function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators({
      ...BlockchainsActions,
      ...SettingsActions,
      ...ValidateActions,
    }, dispatch)
  };
}

export default compose(
  withTranslation('tools', {
    withRef: true
  }),
  connect(mapStateToProps, mapDispatchToProps)
)(GlobalBlockchainForm);
