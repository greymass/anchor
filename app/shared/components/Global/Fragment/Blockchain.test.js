import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import { GlobalFragmentBlockchain } from './Blockchain';

const mockProps = {
  blockchain: {},
  t: jest.fn,
};

describe('GlobalFragmentBlockchain', () => {
  it('renders correctly', () => {
    const wrapper = shallow(<GlobalFragmentBlockchain {...mockProps} />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
