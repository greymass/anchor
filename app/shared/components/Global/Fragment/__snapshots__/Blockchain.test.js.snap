// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GlobalFragmentBlockchain renders correctly 1`] = `
<Popup
  content="Chain ID: undefined"
  hoverable={true}
  inverted={true}
  keepInViewPort={true}
  on="hover"
  position="top left"
  trigger={<span />}
/>
`;
